coverage/docs/design.md

```diff
@@ -0,0 +1,113 @@
+
+##Pre-submit workflow
+Runs code coverage tool to report coverage change in a new commit or new PR
+1. Developer submit new commit to an open PR on github
+2. Matching pre-submit prow job is started 
+3. Test coverage profile generated
+4. Calculate coverage changes. Compare the coverage file generated in this cycle against the most recent successful post-submit build. Coverage file for post-submit commits were generated in post-submit workflow
+5. Use PR data from github, git-attributes, as well as coverage change data calculated above, to produce a list of files that we care about in the line-by-line coverage report. produce line by line coverage html and add link to covbot report. Note that covbot is the robot github account used to report code coverage change results.
+6. Let covbot post presubmit coverage on github, under that conversation of the PR. 
+7. When coverage threshold is enforced, block PR from merging by making this prow job 'required' and return with a code other than 0
+
+##Post-submit workflow
+Produces & stores coverage profile for later presubmit jobs to compare against
+1. A PR is merged
+2. Post-submit prow job started
+3. Test coverage profile generated. Completion marker generated upon successful run. Both stored as prow artifacts.
+    - Completion marker is used by later pre-submit job when searching for a healthy and complete code coverage profile in the post-submit jobs
+
+##Periodical workflow
+Produces periodical coverage result as input for TestGrid
+1. Periodical prow job starts periodically based on the specification in prow job config
+2. Generate coverage profile
+3. Generate / store per-file coverage data
+  - Stores in the XML format, that is used by TestGrid, and dump it in prow artifacts directory 
+  
+##Prow Configuration File
+As mentioned earlier, we use configuration file to store repository specific information. Below is an example that contains the args that will be supplied to the coverage container
+```$xslt
+  - name: pull-knative-serving-go-coverage
+    labels:
+      preset-service-account: "true"
+    agent: kubernetes
+    context: pull-knative-serving-go-coverage
+    always_run: true
+    rerun_command: "/test pull-knative-serving-go-coverage"
+    trigger: "(?m)^/test (all|pull-knative-serving-go-coverage),?(\\s+|$)"
+    optional: true
+    decorate: true
+    clone_uri: "git@github.com:knative/serving.git"
+    ssh_key_secrets:
+    - ssh-knative
+    spec:
+      containers:
+      - image: gcr.io/knative-tests/test-infra/coverage:latest
+        imagePullPolicy: Always
+        command:
+        - "/coverage"
+        args:
+        - "--postsubmit-gcs-bucket=knative-prow"
+        - "--postsubmit-job-name=post-knative-serving-go-coverage"
+        - "--artifacts=$(ARTIFACTS)" # folder to store artifacts, such as the coverage profile
+        - "--profile-name=coverage_profile.txt"
+        - "--cov-target=./pkg/" # target directory of test coverage
```

Comment:
Can we use `go list` here? Do we expect this tool to get everything or just specific directories?

Reply:
just specific directory. We should allow each repo to specify the directory they care about for code coverage

Reply:
just the directories supplied here